src/gc/gc.cpp

```diff
@@ -33443,17 +33809,58 @@ HRESULT GCHeap::Initialize ()
 #endif //SEG_MAPPING_TABLE
 
 #ifdef MULTIPLE_HEAPS
+    uint32_t nhp_from_config = static_cast<uint32_t>(GCConfig::GetHeapCount());
+
+    // GetGCProcessCpuCount only returns up to 64 procs.
+    unsigned int nhp_from_process = GCToOSInterface::CanEnableGCCPUGroups() ?
+                                    GCToOSInterface::GetTotalProcessorCount():
+                                    GCToOSInterface::GetCurrentProcessCpuCount();
+
+    unsigned int nhp = ((nhp_from_config == 0) ? nhp_from_process :
+                                                 (min (nhp_from_config, nhp_from_process)));
+
+
+    nhp = min (nhp, MAX_SUPPORTED_CPUS);
+
     if (GCConfig::GetNoAffinitize())
         gc_heap::gc_thread_no_affinitize_p = true;
 
-    uint32_t nhp_from_config = static_cast<uint32_t>(GCConfig::GetHeapCount());
-    
-    uint32_t nhp_from_process = GCToOSInterface::GetCurrentProcessCpuCount();
+#if !defined(FEATURE_REDHAWK) && !defined(FEATURE_CORECLR)
+    if (!(gc_heap::gc_thread_no_affinitize_p))
+    {
+        if (!(GCToOSInterface::CanEnableGCCPUGroups()))
+        {
+            size_t gc_thread_affinity_mask = static_cast<size_t>(GCConfig::GetGCHeapAffinitizeMask());
 
-    uint32_t nhp = ((nhp_from_config == 0) ? nhp_from_process :
-                                             (min (nhp_from_config, nhp_from_process)));
+            uintptr_t pmask, smask;
+            if (GCToOSInterface::GetCurrentProcessAffinityMask(&pmask, &smask))
+            {
+                pmask &= smask;
 
-    nhp = min (nhp, MAX_SUPPORTED_CPUS);
+                if (gc_thread_affinity_mask)
+                {
+                    pmask &= gc_thread_affinity_mask;
+                }
+
+                process_mask = pmask;
+
+                unsigned int set_bits_in_pmask = 0;
+                while (pmask)
```

Comment:
A nit - an alternative if-less way to get number of set bits would be
```c++
while (pmask != 0)
{
    set_bits_in_pmask++;
    pmask &= (pmask - 1);
}
```

Reply:
yep that's more efficient - but it makes no actual difference cause other calls will dominate so I'll leave it the same as on desktop (code is a bit easier to read).